Update firebase quickstart (#856)
clerk/clerk-docs: docs/integrations/databases/firebase.mdx

# Integrate Firebase with Clerk

<TutorialHero
  beforeYouStart={[
    {
      title: "Set up a Clerk application",
      link: "/docs/quickstarts/setup-clerk",
    },
    {

>       title: "Set up a local project",

Author: Good idea

      link: "/docs/quickstarts/overview",

>     },

Reviewer: add the "Set up a Firebase application"

  ]}
>

- Use Clerk to authenticate access to your Firebase data.
- Configure your Firebase Clerk integration.

</TutorialHero>

> Integrating Firebase with Clerk gives you the benefits of using Firebase's NoSQL database while leveraging Clerk's authentication, prebuilt components, and webhooks.

Reviewer: ```suggestion
Integrating Firebase with Clerk gives you the benefits of using Firebase's NoSQL database while leveraging Clerk's authentication, prebuilt components, and webhooks.
```

Author: I think you forgot to add a change here

<Steps>

### Configure the integration 

To get started:
1. Navigate to the Clerk Dashboard.
1. In the sidebar, select **[Integrations](https://dashboard.clerk.com/last-active?path=integrations)**, and toggle the Firebase integration on.

The Firebase integration enables you to use Clerk to generate a valid authentication token to send to Firebase's API. To do so, you can upload your Firebase project's service account key JSON file to configure the integration automatically.

To generate a service account key:

1. Visit [the Service Accounts settings](https://console.firebase.google.com/project/_/settings/serviceaccounts/adminsdk) in your Firebase project.
1. Near the bottom of the page, select **Generate New Private Key** and save the file.

See [the manual configuration instructions](#configure-firebase-manually) if you don't want to use a service account key.

### Enable authentication in Firebase

To use Firebase auth, ensure authentication is enabled in your Firebase dashboard. To do so:

> 1. Navigate to your Firebase dashboard.
> 1. In the sidebar navigation, expand **Build** and select **Authentication**.
> 1. Enable authentication on this page, and enable any sign-in method you want, such as **Email/Password**.

Reviewer: Does the sign-in method enabled here need to enabled in the Clerk Dashboard as well? So if they enable twitter, does twitter need to be enabled in Clerk as well?

```suggestion
1. Navigate to your Firebase dashboard.
1. In the navigation sidebar, select the **Build** dropdown and select [**Authentication**](https://console.firebase.google.com/u/0/project/_/authentication).
2. Select **Get started**.
3. Enable any sign-in method you want, such as **Email/Password**.
```

> ### Add a Security Rule to your Firestore database

Reviewer: Do all users need to do this step?

```suggestion
### Add a Security Rule to your Firestore database (optional)
```

Reviewer: okay just got to the last step, and I see that the example includes firestore. so we should make it clear in this section that for the sake of this tutorial, the user will need to setup firestore.
I've never set up in the firebase dashboard or used it in a project, so i'm not sure what that entails.

To use Firestore with Clerk, ensure that you have defined [Security Rules](https://firebase.google.com/docs/firestore/security/get-started) that allow authenticated requests to access your database. For example:

```bash
service cloud.firestore {
  match /databases/{database}/documents {
    match /{document=**} {
      allow read, write: if request.auth != null;
    }
  }
}
```

### Get your Firebase config object

To connect to your Firebase app in your code, you need a config object from your Firebase project. To find it:

1. Visit [your Firebase project settings](https://console.firebase.google.com/project/_/settings/general/).

> 1. Scroll down to the setup code snippet on this page, and copy the config. It should be similar to the following:

Reviewer: ```suggestion
1. In the **Your apps** section, there should be a code snippet that includes the `firebaseConfig` object. Copy this object. It should look similar to the following:
```

    ```typescript
    const firebaseConfig = {
      apiKey: "xxxxxxxxxxxxxxxxxxxxxxxxxxxxxx",
      authDomain: "clerk-example-xxxxx.firebaseapp.com",
      databaseURL: "https://clerk-example-xxxxx-default-xxxx.firebaseio.com",
      projectId: "clerk-test-xxxx",
      storageBucket: "clerk-test-xxxx.appspot.com",
      messagingSenderId: "012345678910",
      appId: "1:012345678:web:abcdef123456hijklm",
      measurementId: "G-ABC123DEF"
    };
    ```
1. Save this information somewhere secure. You'll need it to connect to your Firebase app.

See [Google's Firebase documentation](https://support.google.com/firebase/answer/7015592) for more information on the config object.

> ### Use Firebase with Clerk in your code

Reviewer: ```suggestion
### Use Firebase with Clerk in your code

Now that you have configured the integration, and you have retrieved your Firebase config object, it's time to use Firebase with Clerk in your code.
```

The following example:
- Uses Clerk to generate an authentication token for Firebase's API.

> - Creates a button for signing into your Firebase app.

Reviewer: Pending the conversation we had - not sure how the auth thing works with firebase auth *and* clerk auth

```suggestion
- Creates a button for signing into your Firebase app using Firebase auth and Clerk auth.
```

> - Creates a button for fetching example data from your Firestore database.

Reviewer: ```suggestion
- Creates a button for fetching example data from your Firestore database.

This example is written for Next.js App Router, but is supported by any React meta framework, such as Remix or Gatsby.
```

```tsx filename="app/firebase/page.tsx"
"use client";
import { useAuth } from "@clerk/nextjs";
import { initializeApp } from "firebase/app";
import { getAuth, signInWithCustomToken } from "firebase/auth";
import { getFirestore } from "firebase/firestore";
import { doc, getDoc } from "firebase/firestore";

const firebaseConfig = {
  apiKey: "xxxxxxxxxxxxxxxxxxxxxxxxxxxxxx",
  authDomain: "clerk-example-xxxxx.firebaseapp.com",
  databaseURL: "https://clerk-example-xxxxx-default-xxxx.firebaseio.com",
  projectId: "clerk-test-xxxx",
  storageBucket: "clerk-test-xxxx.appspot.com",
  messagingSenderId: "012345678910",
  appId: "1:012345678:web:abcdef123456hijklm",
  measurementId: "G-ABC123DEF",
};

// Connect to your Firebase app
const app = initializeApp(firebaseConfig);
// Connect to your Firestore database
const db = getFirestore(app);

> const auth = getAuth(app);

Reviewer: ```suggestion
// Connect to Firebase auth
const auth = getAuth(app);
```

- Creates a button for signing into your Firebase app.
- Creates a button for fetching example data from your Firestore database.

This example is written for Next.js App Router, but is supported by any React meta framework, such as Remix or Gatsby.

```tsx filename="app/firebase/page.tsx"
"use client";
import { useAuth } from "@clerk/nextjs";
import { initializeApp } from "firebase/app";
import { getAuth, signInWithCustomToken } from "firebase/auth";
import { getFirestore } from "firebase/firestore";
import { doc, getDoc } from "firebase/firestore";

const firebaseConfig = {
  apiKey: "xxxxxxxxxxxxxxxxxxxxxxxxxxxxxx",
  authDomain: "clerk-example-xxxxx.firebaseapp.com",
  databaseURL: "https://clerk-example-xxxxx-default-xxxx.firebaseio.com",
  projectId: "clerk-test-xxxx",
  storageBucket: "clerk-test-xxxx.appspot.com",
  messagingSenderId: "012345678910",
  appId: "1:012345678:web:abcdef123456hijklm",
  measurementId: "G-ABC123DEF",
};

// Connect to your Firebase app
const app = initializeApp(firebaseConfig);
// Connect to your Firestore database
const db = getFirestore(app);
// Connect to Firebase auth
const auth = getAuth(app);

> const getFirestoreData = async () => {

Reviewer: ```suggestion
// Remove this if you do not have Firestore set up
// for your Firebase app
const getFirestoreData = async () => {
```

3. Toggle the **Firebase** integration on. The configuration modal will appear. Keep this open while you configure your Firebase project.

To configure the integration, you can either:
- [upload your Firebase project's service account key JSON file to configure the integration automatically](#upload-service-account-key)
- [manually provide the required information about your Firebase project](#configure-firebase-manually)

#### Upload service account key

To generate a service account key:

1. In your Firebase project, visit [the Service Accounts settings](https://console.firebase.google.com/project/_/settings/serviceaccounts/adminsdk).
2. Near the bottom of the page, select the **Generate new private key** button.
3. In the modal that pops up, select the **Generate key** button to download the JSON file that contains your service account key.
4. In the Clerk Dashboard, the Firebase configuration modal should still be open. Select the **Upload service account key" button and upload the JSON file you downloaded.
5. The appropriate fields in the configuration modal will be filled in automatically. Select **Apply changes** to save your configuration.

> See [the manual configuration instructions](#configure-firebase-manually) if you don't want to use a service account key.

Author: I don't want it to be a step, because it's an alternative rather than a necessary step. Instead, I put the content in tabs

> See [the manual configuration instructions](#configure-firebase-manually) if you don't want to use a service account key.

Reviewer: yes, it wouldn't be a step, it would have been an h4 within the h3 step. but the tabs is good too !

</TutorialHero>

Integrating Firebase with Clerk gives you the benefits of using Firebase's features while leveraging Clerk's authentication, prebuilt components, and webhooks.

<Steps>

### Configure the integration 

> The Firebase integration enables you to use Clerk to generate a valid authentication token to send to Firebase Auth. This enables you to leverage Clerk's prebuilt components, auth provider options, and more, while accessing Firebase products like Firestore with a session validated by Firebase Auth.

Reviewer: Nice! I do like it better here, good idea
